pkg/replica/hash.go

```diff
@@ -0,0 +1,413 @@
+	fileInfo, err := os.Stat(diskutil.GenerateSnapshotDiskName(snapshotName))
+	if err != nil {
+		return "", err
+	}
+
+	return fileInfo.ModTime().String(), nil
+}
+
+func SetSilentlyCorruptedToXattr(snapshotName string) error {
+	info, err := GetSnapshotHashInfoFromXattr(snapshotName)
+	if err != nil && err != syscall.ENODATA {
+		return err
+	}
+
+	if info == nil {
+		info = &SnapshotXattrHashInfo{
+			Method:       defaultHashMethod,
+			Checksum:     info.Checksum,
+			ModTime:      info.ModTime,
+			LastHashedAt: info.LastHashedAt,
+		}
+	}
+
+	info.SilentlyCorrupted = true
+
+	return SetSnapshotHashInfoToXattr(snapshotName, info)
+}
+
+func GetSnapshotHashInfoFromXattr(snapshotName string) (*SnapshotXattrHashInfo, error) {
+	xattrSnapshotHashValue := make([]byte, xattrSnapshotHashValueMax)
+	_, err := unix.Getxattr(diskutil.GenerateSnapshotDiskName(snapshotName), xattrSnapshotHashName, xattrSnapshotHashValue)
+	if err != nil {
+		return nil, err
+	}
+
+	index := bytes.IndexByte(xattrSnapshotHashValue, 0)
+
+	info := &SnapshotXattrHashInfo{}
+	if err := json.Unmarshal(xattrSnapshotHashValue[:index], info); err != nil {
+		return nil, err
+	}
+
+	return info, nil
+}
+
+func SetSnapshotHashInfoToXattr(snapshotName string, info *SnapshotXattrHashInfo) error {
+	xattrSnapshotHashValue, err := json.Marshal(&SnapshotXattrHashInfo{
+		Method:            defaultHashMethod,
+		Checksum:          info.Checksum,
+		ModTime:           info.ModTime,
+		LastHashedAt:      info.LastHashedAt,
+		SilentlyCorrupted: info.SilentlyCorrupted,
+	})
+	if err != nil {
+		return err
+	}
+
+	return unix.Setxattr(diskutil.GenerateSnapshotDiskName(snapshotName), xattrSnapshotHashName, xattrSnapshotHashValue, 0)
+}
+
+func DeleteSnapshotHashInfoFromXattr(snapshotName string) error {
+	return unix.Removexattr(diskutil.GenerateSnapshotDiskName(snapshotName), xattrSnapshotHashName)
+}
+
+func (t *SnapshotHashJob) isSilentCorruptionAlreadyDetected(currentModTime string) (bool, error) {
+	info, err := GetSnapshotHashInfoFromXattr(t.SnapshotName)
+	if err != nil || info == nil {
+		if err != syscall.ENODATA {
+			return false, errors.Wrapf(err, "failed to get snapshot %v last hash info from xattr", t.SnapshotName)
+		}
+		return false, nil
+	}
+
+	if currentModTime == info.ModTime {
+		return info.SilentlyCorrupted, nil
+	}
+
+	return false, nil
+}
+
+func (t *SnapshotHashJob) isRehashRequired(currentModTime string) (bool, string, error) {
+	info, err := GetSnapshotHashInfoFromXattr(t.SnapshotName)
+	if err != nil || info == nil {
+		if err != syscall.ENODATA {
+			return true, "", errors.Wrapf(err, "failed to get snapshot %v last hash info from xattr", t.SnapshotName)
+		}
+		return true, "", nil
+	}
+
+	checksum := info.Checksum
+	modTime := info.ModTime
+
+	if modTime != currentModTime || checksum == "" {
+		return true, "", nil
+	}
+
+	return false, checksum, nil
+}
+
+func (t *SnapshotHashJob) isModTimeRemain(oldModTime string) error {
```

Comment:
In general, `isXXX` is for returning bool, but of course can have an extra error returned. Here, I would suggest you return (bool, error), but change the code to below.

```go
newModTime, err := GetSnapshotModTime(t.SnapshotName)
if err != nil {
    return false, err
}

return oldModTime != newModTime, nil
```